src/compute/src/render/join/linear_join.rs

```diff
@@ -88,18 +91,30 @@ impl LinearJoinSpec {
         V2: Data,
     {
         use LinearJoinImpl::*;
-        use YieldSpec::*;
 
-        match (self.implementation, self.yielding) {
-            (DifferentialDataflow, _) => {
+        match (
+            self.implementation,
+            self.yielding.after_work,
+            self.yielding.after_time,
+        ) {
+            (DifferentialDataflow, _, _) => {
                 differential_dataflow::operators::JoinCore::join_core(arranged1, arranged2, result)
             }
-            (Materialize, ByWork(limit)) => {
-                let yield_fn = move |_start, work| work >= limit;
+            (Materialize, Some(work_limit), Some(time_limit)) => {
+                let yield_fn =
+                    move |start: Instant, work| work >= work_limit || start.elapsed() >= time_limit;
+                mz_join_core(arranged1, arranged2, shutdown_token, result, yield_fn)
+            }
+            (Materialize, Some(work_limit), None) => {
+                let yield_fn = move |_start, work| work >= work_limit;
+                mz_join_core(arranged1, arranged2, shutdown_token, result, yield_fn)
+            }
+            (Materialize, None, Some(time_limit)) => {
+                let yield_fn = move |start: Instant, _work| start.elapsed() >= time_limit;
```

Comment:
I wonder what the overhead is to checking the option inside the closure. Calling `mz_join_core` several times might be bad for compile times, but checking the limits in the closure might consume more CPU? Unclear!

Reply:
So there are several things we could do:

1. Have a single `yield_fn` that always checks both work and time and handles the `Option`s. Good for readability, good for compile times, bad for runtime performance.
2. Have multiple `yield_fn`s for each `Some`/`None` combination, instantiate a different `mz_join_core` variant for each of them. Bad for readability, bad for compile times, good for runtime performance.
3. Have multiple `yield_fn`s for each `Some`/`None` combination, instantiate a single `mz_join_core` that takes a `Box<dyn YFn>`. Okay for readability, good for compile times, okay for runtime performance.

(The "X for runtime performance" are just guesses, I don't really know either.)

I went with (2) here because I really want to avoid regressing join performance, but I'm not too happy about readability. I'm not too worried about compile times, though maybe I should be :)